### DISCUSSION TITLE
Re-run query/update cache upon running a mutation

### DISCUSSION THREAD

**Discussion by vinaybedre** [urql-graphql/urql#2733] [Q&A]
Hi,

I use https://www.the-guild.dev/graphql/codegen to generate graphql client code with a server URL and local .graphql files.

With this in place, when using Apollo GraphQL, rerunning a query when a mutation is run is something we can decide locally. For eg. amount 50 different queries, 30 different mutations, and 1 client. I can choose, when a mutation is run I can provide list of queries to refetch/invalidate the corresponding cache.

With https://formidable.com/open-source/urql/docs/graphcache/ what I understand from documentation is, at the client level (which will be just 1 for entire app), I need to write code to invalidate cache. This becomes problematic, when you use graphql codegen, and have hundreds of mutations/queries. I also knows that urql automatically fetches updated values, but my point is about adding new entry/removing an entry from cache.

I am strongly considering moving to urql, from Apollo GraphQL, but this is something a blocker for me. I hope some experts can add in their thoughts.

As I am reading more of documentation at https://formidable.com/open-source/urql/docs/graphcache/cache-updates/#cache-updates-outside-updaters I know its not possible to call cache update from outside. But I still do not understand why the design is like this? When I have hundreds of mutations, is it good to define all the updateCache methods inside cacheExchange's update method?

**Comment by JoviDeCroock** [comment]
Hey and welcome,

One of the important things to consider is that you probably won't ever reach a 100 updaters for 100 mutations, as you say yourself it's about linking and unlinking entities rather than anything else. To clarify, our default document-based cache exhibits the behaviour your describe, once you do a mutation it will re-fetch the related queries, graphcache can actually do a similar thing when you use `cache.invalidate` from the methods you want.

The choice for only updating the cache as part of GraphQL operations is very conscious as it removes a big footgun, when we approach the concept of GraphQL we know that data can only change in two cases

- the user performs a mutation and the data updates
- the user is notified through a subscription that the data is modified

our logic accounts for updates happening only for these.

**Reply by vinaybedre** [reply]
Hey @JoviDeCroock 
Thanks for the swift reply.
I checked https://formidable.com/open-source/urql/docs/graphcache/cache-updates/ for `cache.invalidate`
with that given example
```
cacheExchange({
  updates: {
    Mutation: {
      mutationOne(_result, args, cache, _info) {
        cache.invalidate({
          __typename: 'Todo',
          id: args.id,
        });
      },
....
      mutationFifty(_result, args, cache, _info) {
        cache.invalidate({
          __typename: 'Todo',
          id: args.id,
        });
      },
    },
  },
});
```
this is the only single place where I can invalidate, and I wont be getting any reference to `cache` object say in my `useMutation` methods is it? 

The problem I see with this is, if I have 4 developers, working on 4 different things, they all have to update the code to `cache.invalidate` in single file and not in respective `useMutation` methods.

What would you suggest me to do in such case?

**Reply by JoviDeCroock** [reply]
I mean, having a central location for your mutations is much easier to maintain imho compared to every `useMutation` call, especially when it comes to for instance multiple means of deletion/creation (i.e. different inputs and outputs) but up to you all. At the end of the day it's an opinion and you all need to use what you prefer.

I can talk to the cases where I have used urql on bigger projects and I have never seen many mutations in the `updates` 😅 but then again it could highly depend on how granular your mutations are i.e. `createTodoWithCompletedFalse` and `createTodoWithCompletedTrue` are two way to express `createTodo` so if you have very granular ways of creation/deletion you might have more updaters..... but generally updating or if the link is visible from the result _does not need an updater_
